### PR TITLE
Feature/bulk workflow manager v2

### PR DESCRIPTION
**Pull request by davidjgonzalez**
Bulk Workflow Manager with

1. AEM workflow support (v1 behavior)
2. Serial Synthetic WF support
2. Parallel Synthetic WF orchestrated via Fast Action Manager

All runners leverage the FAM ThrottledTaskRunner to respect the CPU and Mem usage limits defined within the OSGi service.

Introduces com.adobe.acs.commons.util.QueryHelper to facilitate accepting multiple query languages (querybuilder, xpath, jcr-sql, jcr-sql and list) and normalizing them into a resource result set (List).

Updates to BWM UI to show
- CPU and Mem utilization
- Time taken
- Projected time remaining
- Failures

Note: This uses oak:Unstructured nodes so this will only work on AEM 6+ with Oak (not TarPM).

image

@badvision I made some minor mods to FAM

AEM 6.2 SNAPSHOTS:
* [acs-aem-commons-content-3.1.1-SNAPSHOT-min.zip](https://github.com/Adobe-Consulting-Services/acs-aem-commons/files/374429/acs-aem-commons-content-3.1.1-SNAPSHOT-min.zip)
* [acs-aem-commons-content-3.1.1-SNAPSHOT.zip](https://github.com/Adobe-Consulting-Services/acs-aem-commons/files/374430/acs-aem-commons-content-3.1.1-SNAPSHOT.zip)

### CODE REVIEW

File: bundle/src/main/java/com/adobe/acs/commons/workflow/bulk/execution/impl/Status.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+package com.adobe.acs.commons.workflow.bulk.execution.impl;
```
Comment (bstopp) [inline bundle/src/main/java/com/adobe/acs/commons/workflow/bulk/execution/impl/Status.java]: Shouldn't this be in a public API, not _impl_?

File: bundle/src/main/java/com/adobe/acs/commons/workflow/bulk/execution/impl/SubStatus.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+package com.adobe.acs.commons.workflow.bulk.execution.impl;
```
Comment (bstopp) [inline bundle/src/main/java/com/adobe/acs/commons/workflow/bulk/execution/impl/SubStatus.java]: Shouldn't this be in a public API, not _impl_?

File: bundle/src/main/java/com/adobe/acs/commons/workflow/bulk/execution/impl/servlets/StartServlet.java [hunk 3]
```diff
@@ -66,63 +63,54 @@ protected final void doPost(SlingHttpServletRequest request, SlingHttpServletRes
...
-            map.put(BulkWorkflowEngine.KEY_BATCH_SIZE,
-                    params.optInt(BulkWorkflowEngine.KEY_BATCH_SIZE, BulkWorkflowEngine.DEFAULT_BATCH_SIZE));
-
-            map.put(BulkWorkflowEngine.KEY_INTERVAL,
-                    params.optInt(BulkWorkflowEngine.KEY_INTERVAL, BulkWorkflowEngine.DEFAULT_INTERVAL));
+            final ModifiableValueMap properties = request.getResource().adaptTo(ModifiableValueMap.class);
 
-            map.put(BulkWorkflowEngine.KEY_BATCH_TIMEOUT,
-                    params.optInt(BulkWorkflowEngine.KEY_BATCH_TIMEOUT, BulkWorkflowEngine.DEFAULT_BATCH_TIMEOUT));
+            properties.put("runnerType", params.getString("runnerType"));
```
Comment (bstopp) [inline bundle/src/main/java/com/adobe/acs/commons/workflow/bulk/execution/impl/servlets/StartServlet.java]: What happened to the constants? 😄

**Comment by bstopp** [comment]
What's the policy on _ getAdministrativeResourceResolver_? Is it possible to ship features w/ default/configurable Service manager Users, or is that not available at this time?

Other than the repetition of string constants and the other previously mentioned notes; this looks good. How's it perform?

**Comment by justinedelson** [comment]
@bstopp I don't think we want to attack the resource resolver question quite yet, except for 6.2-only features. Problem being that in 6.0, there's no safe way to configure service resolvers.

**Comment by bstopp** [comment]
Fair enough; do we want to create an open item to review/prepare for it so it's tracked?

**Comment by justinedelson** [comment]
not a bad idea. although a simple grep will do the same thing :)

File: bundle/src/main/java/com/adobe/acs/commons/workflow/bulk/execution/impl/Status.java (hunk 1 above)
Comment (davidjgonzalez) [inline bundle/src/main/java/com/adobe/acs/commons/workflow/bulk/execution/impl/Status.java]: Very little in BWM should be used outside of Commons; maybe to that point it all goes under impl (cant recall if/what the JSPs use; i dont think much)

Comment (justinedelson) [inline bundle/src/main/java/com/adobe/acs/commons/workflow/bulk/execution/impl/Status.java]: yeah, but I think at least `SubStatus` is referenced from an exported class, so it should be also exported.

Comment (justinedelson) [inline bundle/src/main/java/com/adobe/acs/commons/workflow/bulk/execution/impl/Status.java]: and, in fact, running a build shows this:

    [WARNING] Bundle com.adobe.acs:acs-aem-commons-bundle:bundle:2.5.1-SNAPSHOT : Export com.adobe.acs.commons.workflow.bulk.execution,  has 1,  private references [com.adobe.acs.commons.workflow.bulk.execution.impl],
    [WARNING] Bundle com.adobe.acs:acs-aem-commons-bundle:bundle:2.5.1-SNAPSHOT : Export com.adobe.acs.commons.workflow.bulk.execution.model,  has 1,  private references [com.adobe.acs.commons.workflow.bulk.execution.impl],

Comment (davidjgonzalez) [inline bundle/src/main/java/com/adobe/acs/commons/workflow/bulk/execution/impl/Status.java]: ah gotcha, will take a look.

**Pull request merged**